crates/core/src/constraint_system/prove.rs

```diff
@@ -479,86 +479,148 @@ fn make_masked_flush_witnesses<'a, U, Tower>(
 	oracles: &MultilinearOracleSet<FExt<Tower>>,
 	witness: &mut MultilinearExtensionIndex<'a, PackedType<U, FExt<Tower>>>,
 	flush_oracle_ids: &[OracleId],
+	flushes: &[Flush<FExt<Tower>>],
 ) -> Result<(), Error>
 where
 	U: ProverTowerUnderlier<Tower>,
 	Tower: ProverTowerFamily,
 {
+	let zeros = PackedType::<U, FExt<Tower>>::zero();
+	let ones = PackedType::<U, FExt<Tower>>::one();
 	// The function is on the critical path, parallelize.
-	let indices_to_update: Vec<(OracleId, MultilinearWitness<'a, _>)> = flush_oracle_ids
-		.par_iter()
-		.map(|&flush_oracle| match &oracles[flush_oracle].variant {
-			MultilinearPolyVariant::Composite(composite) => {
-				let inner_polys = composite.inner();
+	let indices_to_update: Vec<(OracleId, MultilinearWitness<'a, _>)> =
+		izip!(flush_oracle_ids, flushes)
+			.map(|(&flush_oracle, flush)| match &oracles[flush_oracle].variant {
+				MultilinearPolyVariant::Composite(composite) => {
+					let inner_polys = composite.inner();
+
+					let selectors = inner_polys
+						.iter()
+						.take(flush.selectors.len())
+						.map(|id| witness.get_multilin_poly(*id))
+						.collect::<Result<Vec<_>, _>>()?;
+
+					let n_vars = composite.n_vars();
+					let log_width = <PackedType<U, FExt<Tower>>>::LOG_WIDTH;
+
+					let packed_len = 1 << n_vars.saturating_sub(log_width);
+
+					let inner_c = composite.c();
+
+					let zero_suffixes = selectors
+						.iter()
+						.map(|poly| {
+							if let Some(packed_evals) = poly.packed_evals() {
+								let mut zero_suffix_len = 0;
+
+								for &packed_evals in packed_evals.iter().rev() {
+									if packed_evals != zeros {
+										break;
+									}
+									zero_suffix_len +=
+										1 << (log_width + poly.log_extension_degree());
+								}
+								zero_suffix_len
+							} else {
+								0
+							}
+						})
+						.collect::<Vec<_>>();
+
+					for (zero_suffix, id, poly) in izip!(&zero_suffixes, inner_polys, selectors) {
+						let nonzero_scalars_prefixes = (1 << oracles.n_vars(*id)) - zero_suffix;
 
-				let polys = inner_polys
-					.iter()
-					.map(|id| witness.get_multilin_poly(*id))
-					.collect::<Result<Vec<_>, _>>()?;
+						witness.update_multilin_poly_with_nonzero_scalars_prefixes([(
+							*id,
+							poly,
+							nonzero_scalars_prefixes,
+						)])?;
+					}
 
-				let n_vars = composite.n_vars();
-				let log_width = <PackedType<U, FExt<Tower>>>::LOG_WIDTH;
+					let polys = inner_polys
+						.iter()
+						.map(|id| witness.get_multilin_poly(*id))
+						.collect::<Result<Vec<_>, _>>()?;
 
-				let packed_len = 1 << n_vars.saturating_sub(log_width);
+					let max_packed_zero_suffix =
+						zero_suffixes.into_iter().max().unwrap_or(0) >> log_width;
 
-				let inner_c = composite.c();
+					let max_packed_zero_suffix = max_packed_zero_suffix.min(packed_len);
```

Comment:
Why is this line needed? When is `max_packed_zero_suffix > packed_len`?

Reply:
This is for the case of a zeroed polynomial with `n_vars < P::LOG_WIDTH`,
where `max_packed_zero_suffix` becomes `1 << (P::LOG_WIDTH + poly.log_extension_degree())`,
because, for optimization, I'm comparing packed fields